### DISCUSSION TITLE
Enable MAUI Essentials on WinUI

### DISCUSSION THREAD

**Discussion by Soap-141** [unoplatform/uno#10985] [Q&A] [answered]
How can I use MAUI Essentials with a WinUI project found the default Uno .NET 6 Template?

I've tried using `UseMauiEssentials` in project configuration but it's only working for the Mobile head.
On Windows head, I'm getting a weird error on launch.

![image](https://user-images.githubusercontent.com/57812285/212695323-75bb304e-3684-4552-82cc-fa53a54b3ac1.png)

I've clicked on `Yes` and I installed what was required but it still doesn't work.

I also tried a normal MAUI project where the Windows platform is in the same head as Android and iOS and it was working while using `UseMaui` which includes MAUI Essentials.

**Comment by jeromelaban** [comment]
Why do you need MAUI essentials? We may be able to cover what's missing using WinRT APIs (or maybe it's already covered).

**Reply by Soap-141** [reply]
Basically I'm investigating the options we have to migrate the Uno Application Template that was using Xamarin Essentials. I know for sure that two of the feature that we are using are working with the Uno implementation because I tested them myself in a project. The only negative is that Uno doesn't provide any interface for dependency injection and having an abstraction already made for us (We are using [this](https://github.com/rdavisau/essential-interfaces) that is now compatible with MAUI Essentials) but we could still make them by ourselves.

Needed Features:
- [x] [Network Status](https://github.com/rdavisau/essential-interfaces/blob/main/src/output/Essential.Interfaces/Maui/MauiEssentials.cs#:~:text=public%20interface%20IConnectivity)
- [x] [Application Version](https://github.com/rdavisau/essential-interfaces/blob/main/src/output/Essential.Interfaces/Maui/MauiEssentials.cs#:~:text=public%20interface%20IAppInfo)
- [ ] [Device Information](https://github.com/rdavisau/essential-interfaces/blob/main/src/output/Essential.Interfaces/Maui/MauiEssentials.cs#:~:text=public%20interface%20IDeviceInfo)
- [ ] [Emails](https://github.com/rdavisau/essential-interfaces/blob/main/src/output/Essential.Interfaces/Maui/MauiEssentials.cs#:~:text=public%20interface%20IEmail)
- [x] [Browser](https://github.com/rdavisau/essential-interfaces/blob/main/src/output/Essential.Interfaces/Maui/MauiEssentials.cs#:~:text=public%20interface%20IBrowser)

I checked the one I know exists and that I tested recently enough.

**Reply by jeromelaban** [reply]
Emails and Browser are available, Device Information is also available to some extents, but it may not provide all the information you need. If you can provide additional details about which information you use, it will help. Thanks!

**Reply by Soap-141** [reply]
- OS Version
- [Device Type](https://github.com/xamarin/Essentials/blob/7218ab88f7fbe00ec3379bd54e6c0ce35ffc0c22/Xamarin.Essentials/DeviceInfo/DeviceInfo.shared.cs)
- [Idiom](https://github.com/xamarin/Essentials/blob/7218ab88f7fbe00ec3379bd54e6c0ce35ffc0c22/Xamarin.Essentials/Types/DeviceIdiom.shared.cs)
- Manufacturer
- Model

**Reply by Soap-141** [reply]
I also just noticed that email doesn't support attachment which we would need for sending log files.
https://github.com/unoplatform/uno/issues/2432

**Reply by francoistanguay** [reply]
If there's missing capabilities in the WinUI API, we'll add them to the Uno Toolkit. No reason to use Essentials.

**Reply by Soap-141** [reply]
@francoistanguay Should I create a feature request or you guys already have it in your Backlog?

**Reply by jeromelaban** [reply]
It's already present as #2432.

**Reply by jeanplevesque** [reply]
I understand that Uno wants to support the same features as MAUI Essentials via WinUI (or other) APIs.
However when something isn't already supported by Uno, **shouldn't it be possible to use MAUI Essentials in Uno (WinUI) apps?**
(Just like it used to be possible to use Xamarin.Essentials in Uno (UWP) apps.)

**Reply by francoistanguay** [reply]
Uno targets more platforms than MAUI so doesnt make much sense. You can always bring in MAUI on a per project basis, but what we're saying  is that IF you have non-UI APIs that Essentials support and WinUI doesnt, we'd like to take the opportunity to add them to Uno Toolkit and reduce the gap.

**Reply by carlh98** [reply]
I'm also in need of the OS version, for iOS we're receiving a number that we can't use

**Reply by jeromelaban** [reply]
> I'm also in need of the OS version, for iOS we're receiving a number that we can't use

Which API are you using, specifically?

**Reply by carlh98** [reply]
We're using Windows.System.Profile.AnalyticsInfo.VersionInfo.DeviceFamilyVersion from Microsoft.Windows.SDK.NET

**Reply by francoistanguay** [reply]
Can we have more info? WHat you're getting/ What you need? What is MAUI using?

We could then decide if it maps to a WinUI API or if it should be brought in Uno Toolkit and how it would work across all platforms.

**Reply by carlh98** [reply]
We have an issue on a specific iOS version, we want to receive email with logs from the client to diagnose better.
When use DeviceFamilyVersion, we receive 4503616807370751 for iOS version instead of 16.4.1.
We need Device Information and Emails with attachments https://github.com/unoplatform/uno/issues/2432.

What is the recommend way to receive the correct Device Information?

**Reply by jeromelaban** [reply]
[DeviceFamilyVersion](https://github.com/unoplatform/uno/blob/974a4a22f69a3668a1787dfed973be0c4de722d8/src/Uno.UWP/System/Profile/AnalyticsVersionInfo.iOS.cs#L18) is providing the correct device information, aligned with what UWP returns. It's weird API, I agree, yet you can get the version this way ([or this way](https://github.com/unoplatform/uno/blob/974a4a22f69a3668a1787dfed973be0c4de722d8/src/SamplesApp/UITests.Shared/Windows_System/Profile/AnalyticsInfoTests.xaml.cs#L20-L26)):

```csharp
ulong v = ulong.Parse(AnalyticsInfo.VersionInfo.DeviceFamilyVersion);
ulong major = (v & 0xFFFF000000000000L) >> 48;
ulong minor = (v & 0x0000FFFF00000000L) >> 32;
ulong build = (v & 0x00000000FFFF0000L) >> 16;
ulong revision = (v & 0x000000000000FFFFL);
var version = new Version(major, minor, build, revision);
```

**Reply by carlh98** [reply]
Thanks @jeromelaban, it's what I need for the OS Version.

**Reply by francoistanguay** [reply]
@kazo0  I feel we could add something in Toolkit non-ui for a helperhere.

**Reply by jeromelaban** [reply]
Indeed, when toolkit can be used, it's would good helper to adapt for this C++ oriented API.

**Reply by MCAleckson** [reply]
I'm porting existing IOS and Android UNO apps to .Net7, and need the old Xamarin.Essentials.Track() and VersionTracking.CurrentVersion functions.

Where might I find these now?  Thanks!

**Reply by Soap-141** [reply]
@MCAleckson Add `<UseMauiEssentials>true</UseMauiEssentials>` in your mobile csproj.

**Reply by MCAleckson** [reply]
Thank you, but when I enable MauiEssentials in this solution, the thing won't compile at all.  So I was hoping the UNO team had implemented other means for getting the old Xamarin.Essentials functionality as Jerome suggested in January on this thread.

But if the answer is "MauiEssentials is the UNO-supported means to use VersionTracking and the rest of Xamarin.Essentials functions", then I'll continue to knock my head against why MauiEssentials breaks my app.

Thanks again!

**Reply by Soap-141** [reply]
@MCAleckson Last thing I can suggest is to make sure you have all the required workloads. dotnet workload --list

**Reply by MCAleckson** [reply]
Here are my workloads, what needs to change?  Thanks again.

Installed Workload Id      Manifest Version                            Installation Source
------------------------------------------------------------------------------------------
maui-windows               8.0.0-preview.7.8842/8.0.100-preview.7      VS 17.7.34009.444
maui-maccatalyst           8.0.0-preview.7.8842/8.0.100-preview.7      VS 17.7.34009.444
maccatalyst                16.4.8694-net8-p7/8.0.100-preview.7         VS 17.7.34009.444
maui-ios                   8.0.0-preview.7.8842/8.0.100-preview.7      VS 17.7.34009.444
ios                        16.4.8694-net8-p7/8.0.100-preview.7         VS 17.7.34009.444
maui-android               8.0.0-preview.7.8842/8.0.100-preview.7      VS 17.7.34009.444
android                    34.0.0-preview.7.382/8.0.100-preview.7      VS 17.7.34009.444

Use `dotnet workload search` to find additional workloads to install.

**Reply by jeromelaban** [reply]
Can you specify the build errors you're getting, [best is a binlog](https://platform.uno/docs/articles/uno-builds-troubleshooting.html).

**Reply by MCAleckson** [reply]
Thank you as always Jerome for stepping in here.  Since it seems UseMauiEssentials=true is the way forward, I will reorient my build towards it.  And I'll for sure bring the build errors here.  Thanks!

**Reply by Soap-141** [reply]
Workloads look fine :/

**Comment by marwalsch** [comment]
I believe this issue is to be blamed on the development environment rather than Uno.
[`<UseMauiEssentials/>`](https://github.com/dotnet/maui/issues/6102) works for WinUI/.NET 6 on my end. From what I understand the Maui abstractions are included in the corresponding .NET-workload rather than a Nuget depenndency as with Xamarin.Essentials, hence the necessary flag in the project file. You must not use `<UseMaui/>` however.
Though i recall stumbling into the same error a while ago, I unfortunately don't remember what exactly fixed it for me. I suggest to ensure that the Maui SDK for windows is actually installed.
![image](https://user-images.githubusercontent.com/72074423/214532098-9ff9f4d9-f0de-40b4-bd35-50a76ead98bb.png)
Running Visual Studio 22 on at least 17.4.3 should provide you with a Essentials version equal or greater than 6.0.548.

**Reply by Soap-141** [reply]
Hello I checked and I have everything.

**Reply by marwalsch** [reply]
Then I'm afraid I'm unable to help any further.


But I can think of one last thing to try, though entirely based on speculation: Try re-installing the workloads through Visual Studio and verify it through: `dotnet workload list`. 

```
Installed Workload Id      Manifest Version       Installation Source
---------------------------------------------------------------------
maui-windows               7.0.52/7.0.100         VS 17.4.33213.308
maui-maccatalyst           7.0.52/7.0.100         VS 17.4.33213.308
maccatalyst                16.1.1477/7.0.100      VS 17.4.33213.308
maui-ios                   7.0.52/7.0.100         VS 17.4.33213.308
android                    33.0.4/7.0.100         VS 17.4.33213.308
ios                        16.1.1477/7.0.100      VS 17.4.33213.308
maui-android               7.0.52/7.0.100         VS 17.4.33213.308
```
The Maui.Essentials Framework reference with the required version is pointed to `C:\Users\*user*\.nuget\packages\microsoft.maui.essentials.ref.win\6.0.548` on my machine.

![image](https://user-images.githubusercontent.com/72074423/215806961-6b0b7a69-d9cb-4dd6-b79a-7928658f06e4.png).

**Reply by Soap-141** [reply]
Tried it, ended up updating my entire workload list and breaking my iOS simulator because the updated version require XCode 14.2 that is not compatible with Visual Studio 17.4.4. It worked for longer, had the output telling me that I had to update a package for Essentials to work, I updated it and still got the same error after.

Anyway, Thanks for the help. @marwalsch

**Comment by Soap-141** [comment, marked as answer]
Fixed using .NET 8 and latest workloads, also using `<PackageReference Include="Microsoft.WindowsAppSDK" Version="1.4.231219000" />`.
